sdk/clientcore/core/src/main/java/io/clientcore/core/util/Union.java

```diff
@@ -0,0 +1,285 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * A class that represents a union of types. A type that is one of a finite set of sum types.
+ * This class is used to represent a union of types in a type-safe manner.
+ *
+ * <p><strong>Create an instance</strong></p>
+ *
+ * <!-- src_embed io.clientcore.core.util.union.UnionJavaDocCodeSnippetsBasic -->
+ * <pre>
+ * Union union = Union.ofTypes&#40;String.class, Integer.class&#41;;
+ * </pre>
+ * <!-- end io.clientcore.core.util.union.UnionJavaDocCodeSnippetsBasic -->
+ *
+ * <p><strong>Create an instance from primitives</strong></p>
+ *
+ * <!-- src_embed io.clientcore.core.util.union.UnionJavaDocCodeSnippetsPrimitiveType -->
+ * <pre>
+ * Union unionPrimitives = Union.ofTypes&#40;int.class, double.class&#41;;
+ * </pre>
+ * <!-- end io.clientcore.core.util.union.UnionJavaDocCodeSnippetsPrimitiveType -->
+ *
+ * <p><strong>Create an instance from collections</strong></p>
+ *
+ * <!-- src_embed io.clientcore.core.util.union.UnionJavaDocCodeSnippetsCollectionType -->
+ * <pre>
+ * Union unionCollections = Union.ofTypes&#40;
+ *     new ParameterizedTypeImpl&#40;List.class, String.class&#41;,
+ *     new ParameterizedTypeImpl&#40;List.class, Integer.class&#41;&#41;;
+ * </pre>
+ * <!-- end io.clientcore.core.util.union.UnionJavaDocCodeSnippetsCollectionType -->
+ *
+ * <p><strong>Consume the value of the Union if it is of the expected type</strong></p>
+ *
+ * <!-- src_embed io.clientcore.core.util.union.UnionJavaDocCodeSnippetsSwitch -->
+ * <pre>
+ * Union union = Union.ofTypes&#40;String.class, Integer.class&#41;;
+ * union.setValue&#40;&quot;Hello&quot;&#41;;
+ * switch &#40;union.getValue&#40;&#41;&#41; &#123;
+ *     case String s -&gt; System.out.println&#40;&quot;String value: &quot; + s&#41;;
+ *     case Integer i -&gt; System.out.println&#40;&quot;Integer value: &quot; + i&#41;;
+ *     default -&gt; throw new IllegalArgumentException&#40;
+ *         &quot;Unknown type: &quot; + union.getCurrentType&#40;&#41;.getTypeName&#40;&#41;&#41;;
+ * &#125;
+ * </pre>
+ * <!-- end io.clientcore.core.util.union.UnionJavaDocCodeSnippetsSwitch -->
+ *
+ * or
+ *
+ * <!-- src_embed io.clientcore.core.util.union.UnionJavaDocCodeSnippetsLambda -->
+ * <pre>
+ * Union union = Union.ofTypes&#40;String.class, Integer.class&#41;;
+ * union.setValue&#40;&quot;Hello&quot;&#41;;
+ * union.tryConsume&#40;
+ *     v -&gt; System.out.println&#40;&quot;String value: &quot; + v&#41;, String.class&#41;;
+ * union.tryConsume&#40;
+ *     v -&gt; System.out.println&#40;&quot;Integer value: &quot; + v&#41;, Integer.class&#41;;
+ * </pre>
+ * <!-- end io.clientcore.core.util.union.UnionJavaDocCodeSnippetsLambda -->
+ *
+ */
+public final class Union {
+    private static final ClientLogger LOGGER = new ClientLogger(Union.class);
+
+    private final List<Type> types;
+    private Object value;
+    private Type currentType;
+
+    private Union(Type... types) {
```

Comment:
To protect ourselves from devolving cases we may want to have checks that all the `types` given are `Class` or `instanceof ParameterizedType`. The usage of `Union` can devolve quickly as there are many more `Type` implementations of than those that we don't appear to have support for

Reply:
https://docs.oracle.com/javase/specs/jls/se21/html/jls-4.html

Reply:
We should also support primitive types like int, long, boolean etc. along with ClassType and ParameterizedType.

Reply:
https://github.com/Azure/azure-sdk-for-java/pull/43778/files#diff-f6577340b3c7d27c91e87204096ea117a925e497e29fbf78587e2d115cb8c8f6R21

Primitive already supported

Reply:
> We should also support primitive types like int, long, boolean etc. along with ClassType and ParameterizedType.

Thankfully primitive types and array types are `Class`s